rake_helper.rb

```diff
@@ -0,0 +1,130 @@
+      secret_access_key: aws_secret_access_key
+    }
+    p 'using configuration: ', aws_configuration
+    if configured?
+      @lambda_client = Aws::Lambda::Client.new(aws_configuration)
+      @lambda_config = yaml["deploy"].find {|conf| conf["function_name"].include? travis_branch.downcase}
+    end
+  end
+
+  def configured?
+    access_key_id && secret_access_key
+  end
+
+
+  def update_lambda_configuration
+    unless configured? && lambda_config
+      p 'insufficient configuration'
+      return nil
+    end
+
+
+    vpc_config = lambda_config["vpc"]
+    environment_variables = lambda_config["environment_variables"]&.map {|str| str.split "="}.to_h
+    layers = lambda_config["layers"]
+    updated_lambda_configuration = [
+      :vpc_config,
+      :environment_variables,
+      :layers
+    ]
+      .select {|key| lambda_config[key]}
+      .map {|key| [key, lambda_config[key]]}
+      .to_h
+    p 'updating_function_configuration with: ', updated_lambda_configuration
+    update_configuration_resp = lambda_client.update_function_configuration(updated_lambda_configuration)
+    p 'update_configuration_resp: ', update_configuration_resp
+  end
+
+  def update_event
+    unless lambda_config["events"]
+      p 'no event config'
+      return nil
+    end
+
+    @event = lambda_config["event"]
+    if event.include?("arn")
+      add_event_source
+    elsif event.include?("SCHEDULE_EXPRESSION")
+      add_cron
+    end
+  end
+
+  def add_event_source
+    existing_events = lambda_client.list_event_source_mappings({function_name: function_name}).event_source_mappings
+    if !existing_events.any? { |event| event.event_source_arn == event }
+      event_to_create = {
+        event_source_arn: event,
+        function_name: function_name,
+        bisect_batch_on_function_error: true,
+        starting_position: "LATEST"
+      }
+      p 'creating event: ', event_to_create
+      create_resp = lambda_client.create_event_source_mapping(event_to_create)
+      p 'created: ', create_resp
+    end
+    existing_events.each do |event|
+      if event.event_source_arn != event
+        p 'deleting event with uuid: ', event.uuid, 'and arn: ', event.event_source_arn
+        lambda_client.delete_event_source_mapping({uuid: event.uuid})
+      end
+    end
+  end
+
+  def add_cron
+    events_client = Aws::CloudWatchEvents::Client.new(configuration)
+    schedule_expression = event.split("=")[1]
+    rule_name = "#{function_name}-rule"
+    p 'rule_name: ', rule_name, 'schedule_expression: ', schedule_expression
+    events_client.put_rule(name: rule_name, schedule_expression: schedule_expression)
+    return p 'missing function_name' unless functin_name
+    target_id = "#{function_name}-lambda"
+    p 'getting lambda with function name', function_name, 'target_id', target_id
+    lambda_resp = lambda_client.get_function(function_name: function_name).configuration
+    arn = lambda_resp.function_arn
+    begin
```

Comment:
I think this block could be simplified a bit by checking the exception type in the rescue and moving the `add_permission` call up a bit so it looks like:

```
begin
  policy_resp = ...
  unless policy_resp.policy.include?...
    permission = lambda_client.add_permission...
  else
    p 'lambda already...'
  end
rescue Aws::Lamda::Errors::ResourceNotFoundException
  p 'no policy'
end
```

Just performs the same checks a little more compactly

Reply:
good call

Reply:
Actually this needs to be slightly more complicated since we want to add a policy if none exists, I've added the changes for that.

Reply:
great, looks like a good solution